environs/bootstrap/bootstrap.go

```diff
@@ -230,14 +220,63 @@ func Bootstrap(ctx environs.BootstrapContext, environ environs.Environ, args Boo
 		}
 	}
 
-	ctx.Infof("Bootstrapping model %q", cfg.Name())
-	logger.Debugf("model %q supports service/machine networks: %v", cfg.Name(), supportsNetworking)
-	disableNetworkManagement, _ := cfg.DisableNetworkManagement()
-	logger.Debugf("network management by juju enabled: %v", !disableNetworkManagement)
+	imageMetadata, err := bootstrapImageMetadata(environ,
+		bootstrapSeries,
+		bootstrapArch,
+		args.BootstrapImage,
+		&customImageMetadata,
+	)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// We want to determine a list of valid architectures for which to pick tools and images.
+	// This includes architectures from custom and other available image metadata.
+	architectures := set.NewStrings()
+	if len(customImageMetadata) > 0 {
+		for _, customMetadata := range customImageMetadata {
+			architectures.Add(customMetadata.Arch)
+		}
+	}
+	if len(imageMetadata) > 0 {
+		for _, iMetadata := range imageMetadata {
+			architectures.Add(iMetadata.Arch)
+		}
+	}
+
+	constraintsValidator, err := environ.ConstraintsValidator()
+	if err != nil {
+		return err
+	}
+
+	// IFF there is any image metadata, then and only then do we update
```

Comment:
> I've found that updating validator vocabulary with empty set will misbehave if validator did not contain this vocabulary. An empty set will become valid vocabulary :D and, in our case, no architecture will be considered valid by that validator.

Isn't that exactly what we want? If the provider has no architecture vocab (because it defers to image metadata), and there is no image metadata, then we should not accept any arch?